### PR TITLE
Add tests for CPU metrics

### PR DESCRIPTION
**Pull request by lfitchett**
This adds a test that makes sure cpu is < 100 and that module cpu is < host cpu

### CODE REVIEW

File: test/modules/MetricsValidator/src/tests/ValidateHostRanges.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            var moduleCpu = cpuMetrics.Where(m => m.Tags.TryGetValue("module", out string module) && module != "host").ToList();
+
+            foreach (var hostCpuQuartile in hostCpu)
+            {
+                this.testReporter.Assert($"{hostCpuQuartile.Key} host CPU < 100%", hostCpuQuartile.Value < 100);
+
+                var moduleQuartile = moduleCpu.Where(m => m.Tags["quantile"] == hostCpuQuartile.Key);
+                foreach (var module in moduleQuartile)
+                {
+                    this.testReporter.Assert($"{hostCpuQuartile.Key} {module.Tags["module"]} CPU <= {hostCpuQuartile.Key} host CPU", module.Value <= hostCpuQuartile.Value);
```
Comment (CindyXing) [inline test/modules/MetricsValidator/src/tests/ValidateHostRanges.cs]: I assume the addition of all modules' CPU usage is smaller than the hostcpu usage. 
We could add that as well.

File: test/modules/MetricsValidator/src/TestBase.cs [hunk 2]
```diff
@@ -20,14 +22,42 @@ public abstract class TestBase
         protected IMetricsScraper scraper;
         protected ModuleClient moduleClient;
 
+        protected string TestName
+        {
+            get { return this.GetType().ToString().Split('.').LastOrDefault(); }
```
Comment (damonbarry) [inline test/modules/MetricsValidator/src/TestBase.cs]: Won't `LastOrDefault` return `null` in the default case? And is that what you want?

File: test/modules/MetricsValidator/src/TestBase.cs [hunk 3]
```diff
@@ -20,14 +22,42 @@ public abstract class TestBase
...
             this.scraper = scraper;
             this.moduleClient = moduleClient;
         }
 
-        public abstract Task Start(CancellationToken cancellationToken);
+        public async Task Start(CancellationToken cancellationToken)
+        {
+            log.LogInformation($"Starting test {this.TestName}");
+            Stopwatch stopwatch = new Stopwatch();
+            stopwatch.Start();
```
Comment (damonbarry) [inline test/modules/MetricsValidator/src/TestBase.cs]: I think these two lines can be condensed into a call to `Stopwatch.StartNew()`

File: test/modules/MetricsValidator/src/tests/ValidateDocumentedMetrics.cs [hunk 4]
```diff
@@ -1,6 +1,6 @@
 // Copyright (c) Microsoft. All rights reserved.
 
-namespace MetricsValidator
+namespace MetricsValidator.Tests
```
Comment (damonbarry) [inline test/modules/MetricsValidator/src/tests/ValidateDocumentedMetrics.cs]: Usually in our codebase child namespaces map to folders (i.e., `test/modules/MetricsValidator/src/tests`)...

Comment (damonbarry) [reply test/modules/MetricsValidator/src/tests/ValidateDocumentedMetrics.cs]: Ah, I see you already have that folder, so this file would normally go into that folder if it's in the corresponding namespace

File: test/modules/MetricsValidator/src/tests/ValidateHostRanges.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        void CheckCPU(List<Metric> metrics)
+        {
+            const string cpuMetricName = "edgeAgent_used_cpu_percent";
+            var cpuMetrics = metrics.Where(m => m.Name == cpuMetricName);
+            this.testReporter.Assert($"{cpuMetricName} metric exists", cpuMetrics.Any(), $"Missing {cpuMetricName}");
+
+            var hostCpu = cpuMetrics.Where(m => m.Tags.TryGetValue("module", out string module) && module == "host").ToDictionary(m => m.Tags["quantile"], m => m.Value);
+            this.testReporter.Assert("Host has all quantiles", hostCpu.Count == 6, $"Host had the following quantiles: {string.Join(", ", hostCpu.Keys)}");
```
Comment (damonbarry) [inline test/modules/MetricsValidator/src/tests/ValidateHostRanges.cs]: What is `hostCpu.Count == 6` doing here?

File: test/modules/MetricsValidator/src/tests/ValidateNumberOfMessagesSent.cs [hunk 6]
```diff
@@ -1,6 +1,6 @@
 // Copyright (c) Microsoft. All rights reserved.
 
-namespace MetricsValidator
+namespace MetricsValidator.Tests
```
Comment (damonbarry) [inline test/modules/MetricsValidator/src/tests/ValidateNumberOfMessagesSent.cs]: namespace <==> folder

File: test/modules/MetricsValidator/src/TestBase.cs (hunk 2 above)
Comment (lfitchett) [reply test/modules/MetricsValidator/src/TestBase.cs]: Yea, but I don't think it's possible for split to return an empty array, so it shouldn't ever happen. I could have used First(), but habit.

File: test/modules/MetricsValidator/src/tests/ValidateDocumentedMetrics.cs (hunk 4 above)
Comment (lfitchett) [reply test/modules/MetricsValidator/src/tests/ValidateDocumentedMetrics.cs]: It already is?

File: test/modules/MetricsValidator/src/tests/ValidateNumberOfMessagesSent.cs (hunk 6 above)
Comment (lfitchett) [reply test/modules/MetricsValidator/src/tests/ValidateNumberOfMessagesSent.cs]: Oh, should I have the src in the namespace? All the other test modules ignore the src folder for their namespace.

File: test/modules/MetricsValidator/src/tests/ValidateHostRanges.cs (hunk 5 above)
Comment (lfitchett) [reply test/modules/MetricsValidator/src/tests/ValidateHostRanges.cs]: Our histograms return 6 quantiles (50, 90, 95, 99, 99.9, 99.99)

File: test/modules/MetricsValidator/src/tests/ValidateDocumentedMetrics.cs (hunk 4 above)
Comment (lfitchett) [reply test/modules/MetricsValidator/src/tests/ValidateDocumentedMetrics.cs]: Oh, should I have the src in the namespace? All the other test modules ignore the src folder for their namespace.

File: test/modules/MetricsValidator/src/tests/ValidateHostRanges.cs (hunk 5 above)
Comment (lfitchett) [reply test/modules/MetricsValidator/src/tests/ValidateHostRanges.cs]: I'll add a comment

File: test/modules/MetricsValidator/src/tests/ValidateDocumentedMetrics.cs (hunk 4 above)
Comment (damonbarry) [reply test/modules/MetricsValidator/src/tests/ValidateDocumentedMetrics.cs]: No you don't need src in the namespace

Comment (damonbarry) [reply test/modules/MetricsValidator/src/tests/ValidateDocumentedMetrics.cs]: And yes, this source file is already in the tests/ subfolder. No idea what I was thinking...

File: test/modules/MetricsValidator/src/tests/ValidateNumberOfMessagesSent.cs (hunk 6 above)
Comment (damonbarry) [reply test/modules/MetricsValidator/src/tests/ValidateNumberOfMessagesSent.cs]: Nope, you're good

**Pull request merged**